src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -1392,6 +1398,12 @@ void MarkField (FieldDefinition field, in DependencyInfo reason)
 				throw new ArgumentOutOfRangeException ($"Internal error: unsupported field dependency {reason.Kind}");
 #endif
 
+			if (reason.Kind == DependencyKind.AlreadyMarked) {
+				Debug.Assert (Annotations.IsMarked (field));
+			} else {
+				Annotations.Mark (field, reason);
+			}
+
```

Comment:
Could this be moved after CheckProcessed?

Reply:
It needs to go before CheckProcessed because the tracing can happen for already-processed items.

Reply:
Mark should happen before CheckProcessed, because multiple sources can mark the same field, and all of those dependencies should be traced (which happens whenever we Mark). It was incorrect before, and now the order is enforced by SetProcess which asserts that a processed item should have been marked (pending) first - that's how I caught this.

Reply:
So could we instead of double marking and adding AlreadyMarked special case not to mark it twice? It seems like this logic is needed only for preserved elements. Could we Mark it from preserve only?

Reply:
Not sure I understand the question - the AlreadyMarked check is there to avoid calling Annotations.Mark twice when a different step marks a member through Annotations (not just for typepreserve or preserved methods). But we still need to call Annotations.Mark if the MarkStep.Mark* method is called directly (for example when we mark declaring types).

Reply:
What I meant is that you can simplify this handling by not doing the double-checking with AlreadyMarked and marked_pending at one line and removing it at the next line.

I think replacing the introduced redundant logic should be fixed. 

- For AlreadyMarked logic you could split the methods and call them with no DependencyInfo from `ProcessMarkedPending` and skip all the checks because the flow went through them already.

- For Mark+CheckProcessed logic

You could replace 
```
			Annotations.Mark (assembly, reason);
			if (CheckProcessed (assembly))
				return;
```

code flow with simple insertion to `processed` and not to touch `marked_pending` at all.

Reply:
Thanks for clarifying - I made this optimization for the `Mark`+`SetProcessed` cases, but like I mentioned in https://github.com/mono/linker/pull/1768#discussion_r562110659, I'd rather not change the behavior of existing callsites that use `CheckProcessed` here, same goes for the `AlreadyMarked` pattern. I opened https://github.com/mono/linker/issues/1776 to track this.

Reply:
I understand the perf impact, but it feel fragile to have a "already marked" version and "non marked" version and try to make sure that all callsites are correct. The overall impact should be pretty small (assuming we're not marking 10000s of methods from custom steps).

Reply:
I agree about the "already marked"/"non marked" split - there might be other ways to address https://github.com/mono/linker/issues/1776 without a split. Also FWIW I did a rough perf measurement on helloworld with this change, and it's probably within noise, but if anything it was a slight improvement (maybe because we avoid walking types in Initialize, though I didn't dig into it).